### ISSUE TITLE
Batch Norm Performance NCHW vs NHWC

### ISSUE THREAD

**Issue by EnricoDeg** [NVIDIA/cudnn-frontend#134]
I tried running batch norm on H100 with the same shapes but different layouts (NCHW vs NHWC).

The performance results show that NCHW is significantly slower than NHWC (about a factor of 2) for both bwd and fwd. I profiled the benchmark and it looks like when batch norm is called with NCHW, there is some conversion to NHWC, then the batch norm is computed and then there is a conversion back to NCHW.

I'm wondering if this is the actual cuDNN backend implementation or if this is somehow a limitation of the frontend. Can you help me out to understand a little bit better what's happening there?

**Comment by ahnseunghae** [comment]
Hello, I’m working on implementing NHWC normalization. Could you please share the code for it.

**Comment by EnricoDeg** [comment]
```
    # Fwd cuDNN case
    # input tensors

    x_gpu = torch.randn(n, c, h, w, dtype=input_type, device="cuda")
    if (layout == "NHWC"):
        x_gpu = x_gpu.to(memory_format=torch.channels_last)
    scale_gpu = torch.randn(1, c, 1, 1, device="cuda") if d == 1 else torch.randn(1, c, 1, 1, 1, device="cuda")
    bias_gpu = torch.randn_like(scale_gpu)
    running_mean_gpu = torch.randn_like(scale_gpu)
    running_var_gpu = torch.randn_like(scale_gpu)

    epsilon_cpu = torch.full((1, 1, 1, 1), epsilon_value)
    momentum_cpu = torch.full((1, 1, 1, 1), momentum_value)

    # output tensors
    saved_mean_gpu = torch.empty_like(running_mean_gpu, device="cuda")
    saved_inv_var_gpu = torch.empty_like(running_var_gpu, device="cuda")
    y_gpu = torch.empty_like(x_gpu, dtype=input_type, device="cuda")

    graph = cudnn.pygraph(
        io_data_type=cudnn.datatypes._torch_to_cudnn_data_type_dict[input_type],
        intermediate_data_type=cudnn.datatypes._torch_to_cudnn_data_type_dict[torch.float],
        compute_data_type=cudnn.datatypes._torch_to_cudnn_data_type_dict[torch.float],
        handle=handle,
    )

    x = graph.tensor_like(x_gpu)
    scale = graph.tensor_like(scale_gpu)
    bias = graph.tensor_like(bias_gpu)

    in_running_mean = graph.tensor_like(running_mean_gpu)
    in_running_var = graph.tensor_like(running_var_gpu)
    epsilon = graph.tensor_like(epsilon_cpu)
    momentum = graph.tensor_like(momentum_cpu)

    y, saved_mean, saved_inv_var, out_running_mean, out_running_var = (
        graph.batchnorm(
            name="BN",
            input=x,
            scale=scale,
            bias=bias,
            in_running_mean=in_running_mean,
            in_running_var=in_running_var,
            epsilon=epsilon,
            momentum=momentum,
        )
    )

    y.set_output(True)
    saved_mean.set_output(True).set_data_type(cudnn.datatypes._torch_to_cudnn_data_type_dict[torch.float])
    saved_inv_var.set_output(True).set_data_type(cudnn.datatypes._torch_to_cudnn_data_type_dict[torch.float])
    out_running_mean.set_output(True).set_data_type(cudnn.datatypes._torch_to_cudnn_data_type_dict[torch.float])
    out_running_var.set_output(True).set_data_type(cudnn.datatypes._torch_to_cudnn_data_type_dict[torch.float])

    graph.validate()
    graph.build_operation_graph()
    graph.create_execution_plans([cudnn.heur_mode.B, cudnn.heur_mode.FALLBACK])
    graph.check_support()
    graph.build_plans()

    variant_pack = {
        x: x_gpu,
        scale: scale_gpu,
        bias: bias_gpu,
        in_running_mean: running_mean_gpu,
        in_running_var: running_var_gpu,
        epsilon: epsilon_cpu,
        momentum: momentum_cpu,
        out_running_mean: running_mean_gpu,
        out_running_var: running_var_gpu,
        saved_mean: saved_mean_gpu,
        saved_inv_var: saved_inv_var_gpu,
        y: y_gpu,
    }
    workspace = torch.empty(graph.get_workspace_size(), device="cuda", dtype=torch.uint8)

    # warm-up runs
    for i in range(bench.nwarm):
        graph.execute(
            variant_pack,
            workspace,
            handle=handle,
        )
    torch.cuda.synchronize()

   # bench runs
    with profile(activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA], record_shapes=True) as prof:
        for i in range(bench.niters):
            graph.execute(
                variant_pack,
                workspace,
                handle=handle,
            )
        torch.cuda.synchronize()

    cuda_time = 0
    for result in prof.key_averages(group_by_input_shape=True):
        cuda_time += result.device_time_total
    return cuda_time / bench.niters / 1e3
```
This is what I've done to benchmark batchnorm with cuDNN frontend python API. It's pretty basic. The performance difference is with any shape.

The conversion between NCHW and NHWC is documented here: https://docs.nvidia.com/deeplearning/cudnn/backend/latest/developer/legacy-api.html#conversion-between-nchw-and-nhwc

But I don't understand how this applies to batch norm. Do you use the tensor core for batch norm? Can this conversion be disabled?

**Comment by Anerudhan** [comment]
Hi @EnricoDeg @ahnseunghae ,

Thanks for raising this issue.
Can we understand the use case that you are trying out? Recently our priority has been RMS and Layer norm (Seen in many LLM models). We would love to understand the urgency of your testing scenario.

Thanks

**Comment by EnricoDeg** [comment]
I was just checking performance of batch norm for both layouts and I found out these large differences. It is not related to a specific use case.
